source-oracle/source_oracle/api.py

```diff
@@ -0,0 +1,285 @@
+        }
+    pool = oracledb.create_pool_async(
+        user=config.credentials.username,
+        password=config.credentials.password,
+        dsn=config.address,
+        min=4,
+        max=4,
+        increment=0,
+        session_callback=init_session,
+        cclass="ESTUARY",
+        purity=oracledb.ATTR_PURITY_SELF,
+        **credentials,
+    )
+
+    return pool
+
+
+async def fetch_tables(
+    log: Logger, pool: oracledb.AsyncConnectionPool,
+) -> list[OracleTable]:
+    async with pool.acquire() as conn:
+        with conn.cursor() as c:
+            sql_columns = ','.join([f.alias for (k, f) in OracleTable.model_fields.items()])
+
+            query = f"SELECT {sql_columns} FROM user_tables"
+
+            tables = []
+            await c.execute(query)
+            async for values in c:
+                cols = [col[0] for col in c.description]
+                row = dict(zip(cols, values))
+                tables.append(
+                    OracleTable(**row)
+                )
+
+            return tables
+
+
+async def fetch_columns(
+    log: Logger, pool: oracledb.AsyncConnectionPool,
+) -> list[OracleColumn]:
+    async with pool.acquire() as conn:
+        with conn.cursor() as c:
+            sql_columns = ','.join(["t." + f.alias for (k, f) in OracleColumn.model_fields.items() if f.alias != 'COL_IS_PK' and not f.exclude])
+
+            query = f"""
+            SELECT {sql_columns}, NVL2(c.constraint_type, 1, 0) as COL_IS_PK FROM user_tab_columns t
+                LEFT JOIN (
+                        SELECT c.table_name, c.constraint_type, ac.column_name FROM all_constraints c
+                            INNER JOIN all_cons_columns ac ON (
+                                c.constraint_name = ac.constraint_name
+                                AND c.table_name = ac.table_name
+                                AND c.constraint_type = 'P'
+                            )
+                        ) c
+                ON (t.table_name = c.table_name AND t.column_name = c.column_name)
+            """
+
+            columns = []
+            await c.execute(query)
+            async for values in c:
+                cols = [col[0] for col in c.description]
+                row = dict(zip(cols, values))
+                row = {k: v for (k, v) in row.items() if v is not None}
+                columns.append(OracleColumn(**row))
+
+            return columns
+
+
+async def fetch_page(
+    # Closed over via functools.partial:
+    table: Table,
+    pool: oracledb.AsyncConnectionPool,
+    # Remainder is common.FetchPageFn:
+    log: Logger,
+    page: str | None,
+    cutoff: Tuple[str],
+) -> AsyncGenerator[Document | str, None]:
+    is_first_query = False
+    if page is None:
+        is_first_query = True
+        async with pool.acquire() as conn:
+            with conn.cursor() as c:
+                await c.execute(f"select min(ROWID) from {table.table_name}")
```

Comment:
This does require a full-table scan according to the execution plan. I've changed this code to just use the string `AAAAAAAAAAAAAAAAAA` which is the minimum possible base64 encoded string of length 18 (ROWID is of length 18), so this can serve as the minimum value for us instead of us having to scan for it.